pkg/helm/factory.go

```diff
@@ -0,0 +1,126 @@
+	InstallChartWithValuesFile(ctx context.Context, chart, ociURI, version, kubeconfigFilePath, valuesFilePath string) error
+	InstallChart(ctx context.Context, chart, ociURI, version, kubeconfigFilePath, namespace, valueFilePath string, skipCRDs bool, values []string) error
+	Template(ctx context.Context, ociURI, version, namespace string, values interface{}, kubeVersion string) ([]byte, error)
+	RegistryLogin(ctx context.Context, registry, username, password string) error
+}
+
+// WithRegistryMirror sets up registry mirror for helm.
+func WithRegistryMirror(mirror *registrymirror.RegistryMirror) Opt {
+	return func(c *Config) {
+		c.RegistryMirror = mirror
+	}
+}
+
+// WithInsecure configures helm to skip validating TLS certificates when
+// communicating with the Kubernetes API.
+func WithInsecure() Opt {
+	return func(c *Config) {
+		c.Insecure = true
+	}
+}
+
+// WithProxyConfig sets the proxy configurations on the Config.
+func WithProxyConfig(proxyConfig map[string]string) Opt {
+	return func(c *Config) {
+		c.ProxyConfig = proxyConfig
+	}
+}
+
+// ExecutableBuilder builds the Helm executable and returns it.
+type ExecutableBuilder interface {
+	BuildHelmExecutable(...Opt) Client
+}
+
+// ClientFactory provides a helm client for a cluster.
+type ClientFactory struct {
+	client         client.Client
+	helmClient     Client
+	mu             sync.Mutex
+	builder        ExecutableBuilder
+	registryMirror *registrymirror.RegistryMirror
+}
+
+// NewClientForClusterFactory returns a new helm ClientFactory.
+func NewClientForClusterFactory(client client.Client, builder ExecutableBuilder) *ClientFactory {
+	hf := &ClientFactory{
+		client:  client,
+		builder: builder,
+		mu:      sync.Mutex{},
+	}
+	return hf
+}
+
+// Get returns a new Helm client configured using information from the provided cluster's management cluster.
+func (f *ClientFactory) Get(ctx context.Context, clus *anywherev1.Cluster) (Client, error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	managmentCluster := clus
+
+	var err error
+	if clus.IsManaged() {
+		managmentCluster, err = clusters.FetchManagementEksaCluster(ctx, f.client, clus)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var rUsername, rPassword string
+	if managmentCluster.RegistryAuth() {
+		rUsername, rPassword, err = configcli.ReadCredentialsFromSecret(ctx, f.client)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	r := registrymirror.FromCluster(managmentCluster)
+
+	// There is no need to rebuild the client everytime,
+	// here we check if the registry has changed and rebuild the helm client in that case
+	// Rebuild the Helm client only if the registry has changed
+	if f.helmClient == nil || !reflect.DeepEqual(f.registryMirror, r) {
+		f.helmClient = f.builder.BuildHelmExecutable(WithRegistryMirror(r), WithInsecure())
+		f.registryMirror = r
+	}
+
+	if r != nil && managmentCluster.RegistryAuth() {
```

Comment:
can we avoid the extra logging if the client was already valid?

TBH, i'm not sure if all this complexity is worth it
With the current code, the only thing you avoid is to build a new struct. You still have to read a cluster object, a secret and perform the login call. Is it worth all the stuff to make it thread safe? Or should we simplify this for now and always build a new client? You should be able to optimize this in the future pretty easily since you have the logic pretty well encapsulated in the factory
You could even add a new factory that provides caching.

Reply:
We could avoid by moving the login into the if, but you have a good point. It may not be worth it since we still have to read the cluster object anyway

We can probably revisit if necessary for optimization. Does simplifying and removing the locking mechanism sound good to you?

Reply:
Agreed to remove it offline